### PR TITLE
[WIP] ENH: Fix mmap for fuse.glusterfs .

### PR DESCRIPTION
**Pull request by jusjusjus**
In short, fuse doesn't support `mmap.mmap` with `mmap.MAP_SHARED` which is the default.

In this patch, the OSError is caught, and mmap is invoked with other settings that work on fuse.glusterfs.
However, I'm not sure which settings are appropriate for `mode='c'` , in which case `prot` is not defined and another error occurs.

See [this](https://stackoverflow.com/questions/46839807/mmap-no-such-device) brief discussion of the issue at hand.

### CODE REVIEW

File: numpy/core/memmap.py [hunk 1]
```diff
@@ -261,7 +261,15 @@ def __new__(subtype, filename, dtype=uint8, mode='r+', offset=0,
...
         array_offset = offset - start
-        mm = mmap.mmap(fid.fileno(), bytes, access=acc, offset=start)
+
+        try:
+            mm = mmap.mmap(fid.fileno(), bytes, access=acc, offset=start)
+        except OSError:
+            prot = mmap.PROT_READ if mode == 'r' else mmap.PROT_WRITE
+            mm = mmap.mmap(fid.fileno(), bytes, offset=start,
+                    flags=mmap.MAP_PRIVATE, prot=prot)
+            # `OSError [19]: No such device` if file via fuse.flusterfs
```
Comment (charris) [inline numpy/core/memmap.py]: `fuse.flusterfs`?

**Comment by charris** [comment]
This makes me nervous as the default behavior depends on the file system.  Can you get the desired behavior using `mode='c'` ? That should give you COW, which I think is what you will get with this patch, although the documentation isn't clear on how it interacts in the absence of fork. Note that Windows does not support the flags argument,, which certainly doesn't matter for fuse, but could be checked before hand in any case.

**Comment by jusjusjus** [comment]
@charris I'll check `mode='c'` on monday.  Thanks for the idea.

I'm not sure if `mmap.MAP_PRIVATE` writes have a use case in the numpy scope:  It [is not specified](http://pubs.opengroup.org/onlinepubs/7908799/xsh/mmap.html) that such write operations are retained on the mapped data, the way I understand the doc.  This would be setting the user up for disastrous behavior.

How about we restricted the case to `assert mode='r'`?  For the 'r' case this would be a minimal fix, and for 'w' it's likely not possible to establish full functionality if `mmap.MAP_SHARED` is not implemented.

Exposing `mmap`'s full API to `memmap` as options (in addition to the simplified `mode`) would be another thing I could implement.  What do you think?

**Comment by charris** [comment]
I really don't have much of an opinion as I don't use `memmap` myself and would prefer others to chime in. @rkern Thoughts?

**Comment by juliantaylor** [comment]
It is a bad idea to make the flags of memmap implicitly dependent on the filesystem being used.

But exposing more of the mmap api is a good idea, we already have several issues open for that.

File: numpy/core/memmap.py [hunk 2]
```diff
@@ -17,6 +17,14 @@
     "write":"w+"
     }
 
+mmap_defaults = {
+    'r':{'access':mmap.ACCESS_READ},
```
Comment (charris) [inline numpy/core/memmap.py]: Need to import `mmap`.

**Comment by charris** [comment]
Any luck with `'c'`?

File: numpy/core/memmap.py (hunk 2 above)
Comment (jusjusjus) [reply numpy/core/memmap.py]: Yeah thanks.  I just saw that `mmap` has lazy import and I respect that now.

**Comment by jusjusjus** [comment]
@charris:  You are right.  Reads (and writes) with `'c'` go through.  Strangely.  I guess we can close this issue, if you don't want to continue opening the mmap API.

**Comment by charris** [comment]
I'd rather that opening the `mmap` be a separate PR addressing some of the current issues, which probably need hunting down. If you want to pursue that, go for it. However, I'm glad that the `c` mode does what you need, so closing this.

**Pull request closed without merging**